libcaf_core/src/actor_system_config.cpp

```diff
@@ -367,13 +368,26 @@ std::string actor_system_config::render_pec(uint8_t x, atom_value,
 
 void actor_system_config::extract_config_file_path(string_list& args) {
   static constexpr const char needle[] = "--caf#config-file=";
-  auto last = args.end();
-  auto i = std::find_if(args.begin(), last, [](const std::string& arg) {
-    return arg.compare(0, sizeof(needle) - 1, needle) == 0;
-  });
-  if (i == last)
-    return;
-  auto arg_begin = i->begin() + strlen(needle);
+  static constexpr const char needle_global[] = "--config-file=";
+  const auto last = args.end();
+  size_t len = 0;
+  auto i = args.begin();
+  // Look for match and set len and i.
+  auto check = [&](const char* str) {
+    i = std::find_if(args.begin(), last, [&](const std::string& arg) {
+      return arg.compare(0, strlen(str), str) == 0;
+    });
+    len = strlen(str);
```

Comment:
I find this code hard to follow because of the side effects going on and the two different needle variables. Why is one called `needle` and the other one `needle_global`? Both of them are global constants. Also, we're adding `config-file` as global option but don't do anything with it. This function still invokes the parser by hand rather than using `config_option::parse` and `config_option::store`.

First of all, I think this code would greatly benefit from turning the side effects into a proper algorithm (untested).

```c++
/// Finds ... in range ...
template <class ForwardIterator, class Sentinel>
std::pair<ForwardIterator, string_view>
find_by_long_name(const config_option& x, ForwardIterator first, Sentinel last) {
  auto long_name = x.long_name();
  for (; first != last; ++first) {
    string_view str{*first};
    // Make sure this is a long option starting with "--".
    if (str.compare(0, 2, "--") != 0)
      continue;
    str.remove_prefix(2);
    // Skip optional "caf#" prefix.
    if (str.compare(0, 4, "caf#") != 0)
      str.remove_prefix(4);
    // Make sure we are dealing with the right key.
    if (str.substr(0, long_name.size()) != long_name)
      continue;
    // Make sure the key is followed by an assignment.
    str.remove_prefix(long_name.size());
    if (str.compare(0, 1, "=") != 0)
      continue;
    // Remove leading '=' and return the value.
    str.remove_prefix(1);
    return {first, str};
  }
  return {first, string_view{}};
}
```

This algorithm is general enough to find its way into `config_option.hpp` as a free function. Bonus points for implementing and using [`string_view::starts_with`](https://en.cppreference.com/w/cpp/string/basic_string_view/starts_with) to further improve readability by getting rid of the awkward `compare(...)`. ;)

After separating the algorithm, the remainder of this function becomes very straightforward.

```c++
  auto ptr = custom_options_.qualified_name_lookup("global.config-file");
  CAF_ASSERT(ptr != nullptr);
  string_list::iterator i;
  string_view path;
  std::tie(i, path) = find_by_long_name(*ptr, args.begin(), args.end());
  if (i == args.end())
    return;
  if (path.empty()) {
    args.erase(i);
    return;
  }
  // ... ptr->parse(...) ... ptr->store(...) ...
```